[WebAuthn] Implement PRF extension + hmac-secret

pascoej opened the pull request:
#### a305a458493c0d4b7835f5fee17cc70295221d38
<pre>
[WebAuthn] Implement PRF extension + hmac-secret
<a href="https://rdar.apple.com/113572812">rdar://113572812</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=259934">https://bugs.webkit.org/show_bug.cgi?id=259934</a>

Reviewed by Brent Fulgham.

Implements the PRF extension for WebAuthn to enable deriving cryptographic
secrets from credentials. The prf extension maps to the CTAP hmac-secret
extension, supporting both credential creation and authentication flows with
salt evaluation. This enables use cases like credential-bound encrypted
storage.

Tests: Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/resources/util.js:
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.cpp:
(WebCore::AuthenticatorAssertionResponse::create):
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::create):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorResponse.cpp:
(WebCore::AuthenticatorResponse::extensions const):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.cpp:
(cbor::CBORReader::readWithBytesConsumed):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::encodeMakeCredentialRequestAsCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::parseAuthenticatorDataExtensions):
(fido::readCTAPMakeCredentialResponse):
(fido::readCTAPGetAssertionResponse):
* Source/WebCore/Modules/webauthn/fido/FidoConstants.h:
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::decryptForProtocol):
(fido::pin::TokenResponse::parse):
(fido::pin::HmacSecretRequest::HmacSecretRequest):
(fido::pin::HmacSecretRequest::create):
(fido::pin::HmacSecretResponse::HmacSecretResponse):
(fido::pin::HmacSecretResponse::parse):
(fido::pin::HmacSecretResponse::output const):
* Source/WebCore/Modules/webauthn/fido/Pin.h:
(fido::pin::HmacSecretRequest::sharedKey const):
(fido::pin::HmacSecretRequest::coseKey const):
(fido::pin::HmacSecretRequest::saltEnc const):
(fido::pin::HmacSecretRequest::saltAuth const):
(fido::pin::HmacSecretRequest::protocol const):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientInputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientInputs.mm:
(-[_WKAuthenticationExtensionsClientInputs dealloc]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputs.mm:
(-[_WKAuthenticationExtensionsClientOutputs initWithAppid:]):
(-[_WKAuthenticationExtensionsClientOutputs initWithAppid:prfEnabled:prfFirst:prfSecond:]):
(-[_WKAuthenticationExtensionsClientOutputs prfFirst]):
(-[_WKAuthenticationExtensionsClientOutputs prfSecond]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputsInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(authenticationExtensionsClientInputs):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueMakeCredentialAfterCheckExcludedCredentials):
(WebKit::CtapAuthenticator::continueGetAssertionAfterCheckAllowCredentials):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.h:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretRequestCreate)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretRequestInvalidSalts)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretResponseRoundTrip)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretResponseInvalidSize)):
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructGetAssertionRequestWithHmacSecret)):
* Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp:
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadMakeCredentialResponseWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetAssertionResponseWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetAssertionResponseWithHmacSecret64)):
* Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h:
(TestWebKitAPI::TestData::std::to_array&lt;uint8_t&gt;):

Canonical link: <a href="https://commits.webkit.org/303406@main">https://commits.webkit.org/303406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6903d94ffc4654b85b8b212198d4cc41920327ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4700 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43235 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139722 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84132 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134077 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4461 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101054 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68376 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135153 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118420 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81849 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82942 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142369 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4369 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37119 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109432 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4450 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109613 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27793 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3310 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114692 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57616 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4423 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33070 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4255 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67869 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4514 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4382 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -69,6 +69,32 @@ static bool hasAtLeastFourCodepoints(const String& pin)
+    if (protocol == PINUVAuthProtocol::kPinProtocol2) {
+        // CTAP 2.1 spec 6.5.7: Protocol 2 decrypt
+        // Split ciphertext into IV (first 16 bytes) and ct (remaining bytes)
+        if (ciphertext.size() < 16)
```
Nit: Might be cleaner to make a constant `ivLength = 16` or something in case this length is ever changed in the spec.

It could be use in lines 80-81, too.

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
```diff
@@ -93,9 +93,74 @@ static Vector<uint8_t> getCredentialId(const Vector<uint8_t>& authenticatorData)
+        return std::nullopt;
+
+    uint8_t flags = authData[rpIdHashLength];
+    bool extensionsPresent = flags & 0x80; // ED flag (bit 7)
```
Rather than a comment, would probably be better to make a constant "ED_flag_bit_7" or something.

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -241,22 +259,36 @@ Vector<uint8_t> encodeGetAssertionRequestAsCBOR(const Vector<uint8_t>& hash, con
 
-        cborMap[CBORValue(4)] = CBORValue(WTFMove(extensionsMap));
+        if (!extensionsMap.empty())
+            cborMap[CBORValue(4)] = CBORValue(WTFMove(extensionsMap));
```
This magic number should probably be a `kExtensionsMap` constant (or whatever the right name would be).

brentfulgham commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -140,19 +140,37 @@ Vector<uint8_t> encodeMakeCredentialRequestAsCBOR(const Vector<uint8_t>& hash, c
 
-        cborMap[CBORValue(6)] = CBORValue(WTFMove(extensionsMap));
+        if (!extensionsMap.empty())
+            cborMap[CBORValue(6)] = CBORValue(WTFMove(extensionsMap));
```
Nit: Magic number.

brentfulgham commented on LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
```diff
@@ -350,4 +350,70 @@
         });
     }, "PublicKeyCredential's [[get]] with empty allowCredentials does NOT skip CTAP2 (resident key safety).");
+
+    promise_test(t => {
```
Are these syncs from upstream? If not, please be sure to push upstream to WPT.

brentfulgham reviewed (commented):
Looks good. This code has a lot of magic numbers (not new with your change), but would be nice to make more constants for these.

Please be sure to push these WPT test changes upstream if they are new tests you created. Alternatively, indicate what WPT hash was used to pull these changes from in the Commit log.

pascoej committed:
[WebAuthn] Implement PRF extension + hmac-secret
rdar://113572812
https://bugs.webkit.org/show_bug.cgi?id=259934

Reviewed by Brent Fulgham.

Implements the PRF extension for WebAuthn to enable deriving cryptographic
secrets from credentials. The prf extension maps to the CTAP hmac-secret
extension, supporting both credential creation and authentication flows with
salt evaluation. This enables use cases like credential-bound encrypted
storage.

Tests: Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp
       Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/resources/util.js:
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.cpp:
(WebCore::AuthenticatorAssertionResponse::create):
* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::create):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorResponse.cpp:
(WebCore::AuthenticatorResponse::extensions const):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.cpp:
(cbor::CBORReader::readWithBytesConsumed):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::encodeMakeCredentialRequestAsCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::parseAuthenticatorDataExtensions):
(fido::readCTAPMakeCredentialResponse):
(fido::readCTAPGetAssertionResponse):
* Source/WebCore/Modules/webauthn/fido/FidoConstants.h:
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::decryptForProtocol):
(fido::pin::TokenResponse::parse):
(fido::pin::HmacSecretRequest::HmacSecretRequest):
(fido::pin::HmacSecretRequest::create):
(fido::pin::HmacSecretResponse::HmacSecretResponse):
(fido::pin::HmacSecretResponse::parse):
(fido::pin::HmacSecretResponse::output const):
* Source/WebCore/Modules/webauthn/fido/Pin.h:
(fido::pin::HmacSecretRequest::sharedKey const):
(fido::pin::HmacSecretRequest::coseKey const):
(fido::pin::HmacSecretRequest::saltEnc const):
(fido::pin::HmacSecretRequest::saltAuth const):
(fido::pin::HmacSecretRequest::protocol const):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientInputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientInputs.mm:
(-[_WKAuthenticationExtensionsClientInputs dealloc]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputs.mm:
(-[_WKAuthenticationExtensionsClientOutputs initWithAppid:]):
(-[_WKAuthenticationExtensionsClientOutputs initWithAppid:prfEnabled:prfFirst:prfSecond:]):
(-[_WKAuthenticationExtensionsClientOutputs prfFirst]):
(-[_WKAuthenticationExtensionsClientOutputs prfSecond]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAuthenticationExtensionsClientOutputsInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(authenticationExtensionsClientInputs):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueMakeCredentialAfterCheckExcludedCredentials):
(WebKit::CtapAuthenticator::continueGetAssertionAfterCheckAllowCredentials):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.h:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretRequestCreate)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretRequestInvalidSalts)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretResponseRoundTrip)):
(TestWebKitAPI::TEST(CtapPinTest, TestHmacSecretResponseInvalidSize)):
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructMakeCredentialRequestWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPRequestTest, TestConstructGetAssertionRequestWithHmacSecret)):
* Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp:
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadMakeCredentialResponseWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetAssertionResponseWithHmacSecret)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetAssertionResponseWithHmacSecret64)):
* Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h:
(TestWebKitAPI::TestData::std::to_array<uint8_t>):

Canonical link: https://commits.webkit.org/303406@main

joostd commented on Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
```diff
@@ -93,41 +93,108 @@ static Vector<uint8_t> getCredentialId(const Vector<uint8_t>& authenticatorData)
+        } else if (hmacIt->second.isByteString()) {
+            auto& hmacOutput = hmacIt->second.getByteString();
+
+            RefPtr<ArrayBuffer> first = ArrayBuffer::tryCreate(hmacOutput.span().first(std::min<size_t>(32, hmacOutput.size())));
```
The hmac-secret values are still encrypted, returning them as plaintext will result in incorrect PRF output values.

b-straub replied:
I can confirm this observation. Testing with a FIDO2 security key on Safari 26.4:                                                                                                                                             
                                                                                                                                                                                                                                
  - Safari macOS ↔ Safari iPadOS: PRF outputs match — both platforms produce consistent results.
  - Safari macOS ↔ Chrome macOS: PRF outputs do not match — the values differ for the same credential and salt inputs.                                                                                                          
   
  This is consistent with the hmac-secret output being returned without decryption. Safari-to-Safari works because both sides apply the same code path, producing identical (but still encrypted) values. Cross-browser         
  comparison reveals the issue since Chrome correctly decrypts the hmac-secret response.
                                                                                                                                                                                                                                
  A test page demonstrating the PRF extension is available at https://b-straub.github.io/BlazorPRF/

b-straub replied:
I did investigate in detail and submitted a bug report:

https://bugs.webkit.org/show_bug.cgi?id=311099

wonathanjong replied:
@pascoej @brentfulgham any updates on fixing this one liner ? like @joostd said on forum, It seems this has already been fixed in branch https://github.com/WebKit/WebKit/tree/WebKit-7624.1.16.13.2/
In particular, here:
https://github.com/WebKit/WebKit/blob/dd80a9e46546e49c4a5da9115458077a51220f49/Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp#L151

Any idea when we can expect a fix to land in Safari? This prf extension would enable a slew of productive use-cases for users of safari - specifically allowing for companies to store encrypted data with zero access and users the ability to decrypt with their existing webauthn passkeys (touch/face id). 

Thank you all for the hard work that you do. It is very much appreciated. 🙏🏻
